tools/istio-nftables/pkg/capture/run.go

```diff
@@ -0,0 +1,1026 @@
+func (cfg *NftablesConfigurator) shortCircuitKubeInternalInterface() {
+	for _, internalInterface := range split(cfg.cfg.RerouteVirtualInterfaces) {
+		cfg.ruleBuilder.InsertRule(constants.PreroutingChain, constants.IstioProxyNatTable, 0, "iifname", internalInterface, "return")
+	}
+}
+
+// shortCircuitExcludeInterfaces adds return rules to skip both NAT and mangle table processing
+// for the interfaces listed in ExcludeInterfaces. This is useful when you want to avoid capturing
+// traffic from specific network interfaces.
+func (cfg *NftablesConfigurator) shortCircuitExcludeInterfaces() {
+	for _, excludeInterface := range split(cfg.cfg.ExcludeInterfaces) {
+		cfg.ruleBuilder.AppendRule(
+			constants.PreroutingChain, constants.IstioProxyNatTable, "iifname", excludeInterface, "return")
+		cfg.ruleBuilder.AppendRule(constants.OutputChain, constants.IstioProxyNatTable, "oifname", excludeInterface, "return")
+	}
+	if cfg.cfg.InboundInterceptionMode == "TPROXY" {
+		for _, excludeInterface := range split(cfg.cfg.ExcludeInterfaces) {
+
+			cfg.ruleBuilder.AppendRule(
+				constants.PreroutingChain, constants.IstioProxyMangleTable, "iifname", excludeInterface, "return")
+			cfg.ruleBuilder.AppendRule(constants.OutputChain, constants.IstioProxyMangleTable, "oifname", excludeInterface, "return")
+		}
+	}
+}
+
+// Run is the main function that builds and applies nftables rules based on the provided configuration.
+// It handles exclusion and inclusion logic for inbound and outbound traffic, DNS redirection,
+// owner-based filtering, TPROXY mark handling, and finally applies all the rules.
+func (cfg *NftablesConfigurator) Run() (*knftables.Transaction, error) {
+	// Since OUTBOUND_IP_RANGES_EXCLUDE could carry ipv4 and ipv6 ranges
+	// need to split them in different arrays one for ipv4 and one for ipv6
+	// in order to not to fail
+	ipv4RangesExclude, ipv6RangesExclude, err := cfg.separateV4V6(cfg.cfg.OutboundIPRangesExclude)
+	if err != nil {
+		return nil, err
+	}
+	if ipv4RangesExclude.IsWildcard {
+		return nil, fmt.Errorf("invalid value for OUTBOUND_IP_RANGES_EXCLUDE")
+	}
+
+	ipv4RangesInclude, ipv6RangesInclude, err := cfg.separateV4V6(cfg.cfg.OutboundIPRangesInclude)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectDNS := cfg.cfg.RedirectDNS
+	// How many DNS flags do we have? Three DNS flags! AH AH AH AH
+	if redirectDNS && !cfg.cfg.CaptureAllDNS && len(cfg.cfg.DNSServersV4) == 0 && len(cfg.cfg.DNSServersV6) == 0 {
+		log.Warn("REDIRECT_DNS is set, but CAPTURE_ALL_DNS is false, and no DNS servers provided. DNS capture disabled.")
+		redirectDNS = false
+	}
+
+	cfg.logConfig()
+
+	// Add rules to skip specific interfaces from redirection.
+	cfg.shortCircuitExcludeInterfaces()
+	cfg.shortCircuitKubeInternalInterface()
+
+	// Create a rule for invalid drop in PREROUTING chain in mangle table, so the nftables will drop the out of window packets instead of reset connection .
```

Comment:
I notice that we're still using iptables naming and conventions within nftables; I'm assuming that intentional? Do we think there's a point in time where we may want to migrate to different/simpler tables now that we're leveraging nftables?

Reply:
Ah just noticed the discussion [here](https://github.com/istio/istio/pull/56487#discussion_r2124607280). Makes sense for a first pass, but I am excited at the prospect of simplifying our redirection logic as much as possible